sdk/core/core/src/az_http_request_builder.c

```diff
@@ -3,48 +3,271 @@
+      || az_const_span_is_overlap(az_span_to_const_span(buffer), initial_url)) {
+    return AZ_ERROR_ARG;
+  }
+
+  if (buffer.size < max_url_size) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+
+  AZ_RETURN_IF_FAILED(az_span_set(buffer, '\0'));
+
+  az_span uri_buf = { 0, 0 };
+  AZ_RETURN_IF_FAILED(az_span_copy(
+      (az_span){ .begin = buffer.begin, .size = max_url_size }, initial_url, &uri_buf));
+
+  *p_hrb = (az_http_request_builder){ .buffer = buffer,
+                                      .method_verb = method_verb,
+                                      .url = uri_buf,
+                                      .max_url_size = max_url_size,
+                                      .max_headers = max_headers,
+                                      .retry_headers_start = max_headers,
+                                      .headers_end = 0 };
+
   return AZ_OK;
 }
 
 az_result az_http_request_builder_set_query_parameter(
     az_http_request_builder * const p_hrb,
     az_const_span const name,
     az_const_span const value) {
-  (void)p_hrb;
-  (void)name;
-  (void)value;
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  if (!az_const_span_is_valid(name) || !az_const_span_is_valid(value) || name.size == 0) {
+    return AZ_ERROR_ARG;
+  }
+
+  size_t new_url_size;
+  bool const has_value = value.size > 0;
+  {
+    az_span new_url_span = { .begin = p_hrb->url.begin };
+    size_t const extra_chars_size = sizeof((uint8_t)'?') + (has_value ? sizeof((uint8_t)'=') : 0);
+    size_t const name_and_value_size = name.size + value.size;
+    size_t const appended_size = name_and_value_size + extra_chars_size;
+
+    new_url_size = p_hrb->url.size + appended_size;
+
+    if (name_and_value_size < name.size || name_and_value_size < value.size
+        || appended_size < name_and_value_size || appended_size < extra_chars_size
+        || new_url_size < appended_size || new_url_size < p_hrb->url.size
+        || new_url_size > p_hrb->max_url_size) {
+      return AZ_ERROR_BUFFER_OVERFLOW;
+    }
+
+    new_url_span.size = new_url_size;
+
+    if (az_const_span_is_overlap(az_span_to_const_span(new_url_span), name)
+        || az_const_span_is_overlap(az_span_to_const_span(new_url_span), value)) {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  bool first_parameter = true;
+  for (size_t i = p_hrb->url.size; i > 0;) {
+    --i;
+    if (p_hrb->url.begin[i] == '?') {
+      first_parameter = false;
+    }
+  }
+
+  p_hrb->url.begin[p_hrb->url.size] = first_parameter ? '?' : '&';
+  ++(p_hrb->url.size);
+
+  memcpy(p_hrb->url.begin + p_hrb->url.size, name.begin, name.size);
+  p_hrb->url.size += name.size;
+
+  if (has_value) {
+    p_hrb->url.begin[p_hrb->url.size] = '=';
+    ++(p_hrb->url.size);
+
+    memcpy(p_hrb->url.begin + p_hrb->url.size, value.begin, value.size);
+    p_hrb->url.size += value.size;
+  }
+
+  assert(p_hrb->url.size == new_url_size);
   return AZ_OK;
 }
 
 az_result az_http_request_builder_append_header(
```

Comment:
All this function is supposed to do is add an az_pair (with the name/value) set to the passed-in arguments) to the end of the array and return. I would expect it to be 2-3 lines of code.

Reply:
Done.